pkg/cmd/roachtest/tests/mixed_version_c2c.go

```diff
@@ -0,0 +1,460 @@
+		Arg("{pgurl%s}", cm.c.Range(1, cm.sp.srcNodes)).
+		Flag("warehouses", 10)
+	tpccRunCmd := roachtestutil.NewCommand("./cockroach workload run tpcc").
+		Arg("{pgurl%s}", cm.c.Range(1, cm.sp.srcNodes)).
+		Option("tolerate-errors").
+		Flag("warehouses", 100)
+	cm.workloadStopper = cm.sourceMvt.Workload("tpcc", cm.c.WorkloadNode(), tpccInitCmd, tpccRunCmd)
+}
+
+func (cm *c2cMixed) LatencyHook(ctx context.Context) {
+	cm.destMvt.BackgroundFunc("latency verifier", func(ctx context.Context, l *logger.Logger, r *rand.Rand, h *mixedversion.Helper) error {
+		lv := makeLatencyVerifier("stream-ingestion", 0, cm.sp.maxAcceptedLatency, l,
+			getStreamIngestionJobInfo, nil, true)
+		defer lv.maybeLogLatencyHist()
+		_, db := h.RandomDB(r)
+
+		// The latency verify doesn't need a stopper, as ctx cancellation will stop it.
+		dummyCh := make(chan struct{})
+		if err := lv.pollLatencyUntilJobSucceeds(ctx, db, int(cm.ingestionJobID), time.Second*5, dummyCh); ctx.Err() == nil {
+			// The ctx is cancelled when the background func is successfully stopped,
+			// therefore, don't return a context cancellation error.
+			l.Printf("latency verifier error: %s", err)
+			return err
+		}
+		return nil
+	})
+}
+
+// UpdateHook registers a few mixed version hooks that ensure that the upgrading
+// clusters obey several invariants, which include:
+// - the destination cluster must be the same or at most one major version ahead
+// of the source cluster. This implies that for a given major upgrade, the
+// destination cluster finalizes before the source cluster.
+// - the app tenant must finalize after the system tenant (baked into the mixed
+// version framework).
+//
+// The hooks also conduct the following:
+// - during random destination side mixed version states, the upgrade processes
+// will wait for the PCR replicated time to catch up, validating the stream can
+// advance in a mixed version state.
+// - After the destination has upgraded to its final version, it will issue a
+// cutover command and fingerprint the app tenant key space.
+// - The source will also run a fingerprint command using the same timestamps
+// and ensure the fingerprints are the same.
+func (cm *c2cMixed) UpdateHook(ctx context.Context) {
+	destFinalized := make(chan struct{}, 1)
+
+	cm.destMvt.InMixedVersion("maybe wait for replicated time",
+		func(ctx context.Context, l *logger.Logger, r *rand.Rand, h *mixedversion.Helper) error {
+			if r.Intn(3) == 0 {
+				cm.midUpgradeCatchupMu.Lock()
+				defer cm.midUpgradeCatchupMu.Unlock()
+				return cm.WaitForReplicatedTime(ctx, nowLess30Seconds(), h, r, 5*time.Minute)
+			}
+			return nil
+		})
+
+	cm.sourceMvt.InMixedVersion(
+		"wait for dest to finalize if source is ready to finalize upgrade",
+		func(ctx context.Context, l *logger.Logger, r *rand.Rand, h *mixedversion.Helper) error {
+			cm.midUpgradeCatchupMu.Lock()
+			l.Printf("acquired mid upgrade lock") // nolint:deferunlockcheck
+			cm.midUpgradeCatchupMu.Unlock()
```

Comment:
Yeah, i should write more words here. I'm attempting to prevent a source node from restarting/changing binaries, because that prevents the replication stream from advancing. As you point out, this lock is best effort. My hope is that at most one call to `sourceMvt.InMixedVersion` runs while we wait for the stream to advance in a mixed version state. 

Ideally I could acquire this lock on every binary node restart, but I don't think the mvt runner allows me to.

Reply:
I'm going to merge as is, since i haven't seen this check flake yet.